File: models/checkout_chat.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

const (
	CheckoutChatRoleUser  = "user"
	CheckoutChatRoleAgent = "agent"

	CheckoutChatStatusActive  = "active"
	CheckoutChatStatusDeleted = "deleted"
)

// CheckoutChat represents a checkout-page discount-negotiation conversation.
// One document per session, scoped to a user rather than a try-on room.
type CheckoutChat struct {
	ID     primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	ChatID string             `bson:"chat_id" json:"chat_id"`
	UserID primitive.ObjectID `bson:"user_id" json:"user_id"`

	Title    string                 `bson:"title" json:"title"`
	Messages []CheckoutChatMessage  `bson:"messages" json:"messages"`
	Metadata CheckoutChatMetadata   `bson:"metadata" json:"metadata"`
	Status   string                 `bson:"status" json:"status"`

	CreatedAt time.Time `bson:"created_at" json:"created_at"`
	UpdatedAt time.Time `bson:"updated_at" json:"updated_at"`
}

// CheckoutChatMessage represents a single message in a checkout negotiation chat.
type CheckoutChatMessage struct {
	ID        string    `bson:"id" json:"id"`
	Role      string    `bson:"role" json:"role"`
	Content   string    `bson:"content" json:"content"`
	Timestamp time.Time `bson:"timestamp" json:"timestamp"`

	ToolCall *CheckoutChatToolCall `bson:"tool_call,omitempty" json:"tool_call,omitempty"`

	ModelUsed      string `bson:"model_used,omitempty" json:"model_used,omitempty"`
	ResponseTimeMs int64  `bson:"response_time_ms,omitempty" json:"response_time_ms,omitempty"`
}

// CheckoutChatToolCall is the embedded tool-call record (offer_coupon only).
type CheckoutChatToolCall struct {
	Name      string                 `bson:"name" json:"name"`
	Arguments map[string]interface{} `bson:"arguments" json:"arguments"`
	Result    map[string]interface{} `bson:"result,omitempty" json:"result,omitempty"`
}

// CheckoutChatMetadata stores per-session analytics.
type CheckoutChatMetadata struct {
	TotalMessages  int      `bson:"total_messages" json:"total_messages"`
	UserMessages   int      `bson:"user_messages" json:"user_messages"`
	AgentMessages  int      `bson:"agent_messages" json:"agent_messages"`
	CouponsOffered []string `bson:"coupons_offered,omitempty" json:"coupons_offered,omitempty"`
	FirstMessageAt time.Time `bson:"first_message_at,omitempty" json:"first_message_at,omitempty"`
	LastMessageAt  time.Time `bson:"last_message_at,omitempty" json:"last_message_at,omitempty"`
}

File: front_end/src/types/checkout-chat.ts
/**
 * What a coupon card needs to draw itself. Kept on the message that offered it;
 * whether it is still the live offer is decided against the session's coupon.
 */
export interface MessageCoupon {
  code: string;
  value: number;
  valid_until: string;
}

export type CheckoutChatRole = "user" | "agent";
export type CheckoutChatStatus = "active" | "deleted";

export interface CheckoutChatToolCall {
  name: string;
  arguments: Record<string, unknown>;
  result?: Record<string, unknown>;
}

export interface CheckoutChatMessage {
  id: string;
  role: CheckoutChatRole;
  content: string;
  timestamp: string;
  tool_call?: CheckoutChatToolCall;
  model_used?: string;
  response_time_ms?: number;
}

export interface CheckoutChatMetadata {
  total_messages: number;
  user_messages: number;
  agent_messages: number;
  coupons_offered?: string[];
  first_message_at?: string;
  last_message_at?: string;
}

export interface CheckoutChat {
  id?: string;
  chat_id: string;
  user_id?: string;
  title: string;
  messages: CheckoutChatMessage[];
  metadata: CheckoutChatMetadata;
  status: CheckoutChatStatus;
  created_at: string;
  updated_at: string;
}

export interface CheckoutChatSessionResponse {
  success: boolean;
  chat: CheckoutChat;
}

/** The finished turn the checkout negotiation stream reports in its `done` event. */
export interface CheckoutNegotiationTurn {
  reply?: string;
  coupon?: MessageCoupon;
}

/** A single message in the checkout discount widget's local UI state. */
export interface CheckoutUIMessage {
  role: "user" | "agent" | "agent_streaming";
  content: string;
  coupon?: MessageCoupon;
}
